fbc: init at 1.09.0 & 1.06-darwin-wip20160505

OPNA2608 opened the pull request:
###### Description of changes

Add the [FreeBASIC compiler](https://freebasic.net/).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

OPNA2608 commented on pkgs/development/compilers/fbc/default.nix:
```diff
@@ -0,0 +1,79 @@
+  version = "1.09.0";
+
+  src = fetchzip {
+    url = "https://github.com/freebasic/fbc/releases/download/${version}/FreeBASIC-${version}-source-bootstrap.tar.xz";
```
Not using `fetchFromGitHub` here because the compiler is written in FreeBASIC itself and needs to be bootstrapped by another fbc installation. Those tarballs provide pre-translated source code so it can be built with just a C compiler / assembler on *some* platforms.

`make bootstrap-minimal` builds a native `fbc` from those files, regular `make` re-translates & re-compiles the final `fbc` with the bootstrapped one.

OPNA2608 replied:
Added a short comment about it.

OPNA2608 commented:
I've checked a downstream project and it seems like I forgot that upstream fbc is actually still not functional on Darwin.

![Bildschirmfoto von 2022-06-24 00-21-44](https://user-images.githubusercontent.com/23431373/175424933-7c84a523-4493-40d5-8f7e-c17345cb2b79.png)

OPNA2608 commented:
This is gonna take abit more work still. I remembered that most tests don't get run in the `checkPhase`, so I tried to look into why I'm skipping the other test sets.

`unit-tests` compiles and starts executing (checkFlag `"UNITTEST_RUN_ARGS=--verbose"` can be used to enable output), but eventually just freezes on:

```console
  fbc_tests.udt_wstring_.midstmt
    ascii
```

It's still doing something according to htop, but I've been waiting for quite awhile without any more changes. Not sure what's happening there, maybe I'm just too impatient.

`log-tests` mostly works, but finishes with 3 failures that don't happen on upstream CI:

```console
FAILED LOG - for log-tests -lang fb
./cpp/derived.log:cpp/derived.bmk : RESULT=FAILED
./cpp/call2.log:cpp/call2.bmk : RESULT=FAILED
./cpp/class.log:cpp/class.bmk : RESULT=FAILED
```

[cpp/derived.log](https://github.com/NixOS/nixpkgs/files/8987345/derived.log)
[cpp/call2.log](https://github.com/NixOS/nixpkgs/files/8987346/call2.log)
[cpp/class.log](https://github.com/NixOS/nixpkgs/files/8987347/class.log)


`cpp/derived` is the only one I started looking into so far. Here is the fbc invocation from the log, repeated with the verbose flag for more details:

```console
[nix-shell:/dev/shm]$ fbc -g /tmp/nix-build-fbc-1.09.0.drv-0/source/tests/cpp/derived-fbc.o /tmp/nix-build-fbc-1.09.0.drv-0/source/tests/cpp/derived-cpp.o -x derived-fbc -v
FreeBASIC Compiler - Version 1.09.0 (2022-06-26), built for linux-x86_64 (64bit)
Copyright (C) 2004-2021 The FreeBASIC development team.
target:       linux-x86_64, x86-64, 64bit
backend:      gcc
linking:      ld -m elf_x86_64 -o "derived-fbc" -dynamic-linker /lib64/ld-linux-x86-64.so.2 -T "/nix/store/gnwwhv4i9k3bnc9jpz911cmd2a25nvrs-j7jv8fppd7glnx56r5nv33nsbfdhhnc9-fbc-1.09.0/bin/../lib/freebasic/linux-x86_64/fbextra.x" -L "/nix/store/gnwwhv4i9k3bnc9jpz911cmd2a25nvrs-j7jv8fppd7glnx56r5nv33nsbfdhhnc9-fbc-1.09.0/bin/../lib/freebasic/linux-x86_64" -L "." -L "/nix/store/d7q0qfm12hb59wj63wyjs0hrdhmmapfz-gcc-11.3.0/lib/gcc/x86_64-unknown-linux-gnu/11.3.0" "/nix/store/lyl6nysc3i3aqhj6shizjgj0ibnf1pvg-glibc-2.34-210/lib/crt1.o" "/nix/store/lyl6nysc3i3aqhj6shizjgj0ibnf1pvg-glibc-2.34-210/lib/crti.o" "/nix/store/d7q0qfm12hb59wj63wyjs0hrdhmmapfz-gcc-11.3.0/lib/gcc/x86_64-unknown-linux-gnu/11.3.0/crtbegin.o" "/nix/store/gnwwhv4i9k3bnc9jpz911cmd2a25nvrs-j7jv8fppd7glnx56r5nv33nsbfdhhnc9-fbc-1.09.0/bin/../lib/freebasic/linux-x86_64/fbrt0.o" "/tmp/nix-build-fbc-1.09.0.drv-0/source/tests/cpp/derived-fbc.o" "/tmp/nix-build-fbc-1.09.0.drv-0/source/tests/cpp/derived-cpp.o" "-(" -lstdc++ -lfb -lncurses -lm -ldl -lpthread -lgcc -lgcc_eh -lc "-)" "/nix/store/d7q0qfm12hb59wj63wyjs0hrdhmmapfz-gcc-11.3.0/lib/gcc/x86_64-unknown-linux-gnu/11.3.0/crtend.o" "/nix/store/lyl6nysc3i3aqhj6shizjgj0ibnf1pvg-glibc-2.34-210/lib/crtn.o" 
/nix/store/2b99rpx8zwdjjqkrvk7kqgn9mxhiidjy-binutils-2.38/bin/ld: cannot find -lstdc++: No such file or directory
linking failed: 'ld' terminated with exit code 1
```

- I think `-dynamic-linker /lib64/ld-linux-x86-64.so.2` could be a problem? In a derivation that might get patched at the end of the build anyway but for interactive usage it's not great.
- The compiler seems to need `${gcc-unwrapped.lib}/lib` / `${stdenv.cc.cc}/lib` (or whatever gave the unwrapped one of the stdenv) added *somewhere* to link `-lstdc++` properly. The required linker flag could be added manually with `-Wl` but that is gonna be hackier than figuring out where the linker flags come from & how to add the required one.
- Seeing `2022-06-26` in there makes me sad, I really wanna patch such build dates away…

OPNA2608 commented:
I've opened https://github.com/freebasic/fbc/issues/387 for the build date and hardcoded some values in the mean time.

All tests now work thanks to some upstream fixes (https://github.com/freebasic/fbc/issues/382) and custom changes (https://github.com/freebasic/fbc/issues/385), at least on x86_64-linux.

I've also spent some time with splitting up the build process so `pkgsCross.<target>.fbc` should mostly work for supported platforms (gnu32, ppc64, mingwW64):

- do a minimal bootstrap of the pre-translated sources, for the build platform
- use the bootstrapped version to translate & build another minimal version of the patched one, with the compiler built for the build and the rtlib for the host platform
- do a full rebuild of the sources with a patched version, completely for the host platform now
(some unpatched changes from the bootstrap compiler remain in the first patched build and patching the pre-translated sources seemed trickier to me than this song & dance)

We can also run some of these on ofborg (and hope that the demo effect doesn't backstab me here):

@ofborg build pkgsCross.ppc64.fbc pkgsCross.mingwW64.fbc

`-dynamic-linker` I don't quite know how to tackle. The various linker paths are actually [hardcoded in the compiler based on the target platform](https://github.com/freebasic/fbc/blob/42f4f6dfdaafdd5302a647152f16cda78e4ec904/src/compiler/fbc.bas#L848-L870). Inside of the build environments it shouldn't matter because the path gets patched anyway, just something to keep in mind when used interactively I suppose.

OPNA2608 commented:
Whoops, I didn't mean to retrigger the cross builds by editing the message… :sweat: 

I think this should be okay for review now. I'd especially appreciate some eyes on that cross stuff cuz that's still relatively new territory to me. I also want to look into setting up a cross-fbc at some point: On the building side it should just be some additional `postBuild` & `postInstall` phases to build & ship the targetPlatform's rtlib etc but I haven't looked into what I need to tell Nix w/r/t targetPlatform dependencies… Maybe another time ^^".

I'll try rebasing another branch with an fbc project on this as an in-the-field test when I have more time, but the included tests have me *semi*-confident that I didn't bork anything.

OPNA2608 commented:
> I'll try rebasing another branch with an fbc project on this as an in-the-field test when I have more time

Tested a FreeBASIC project on this branch, still worked fine.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1065

SuperSandro2000 commented on pkgs/development/compilers/fbc/mac-bin.nix:
```diff
@@ -0,0 +1,35 @@
+  version = "1.06-darwin-wip20160505";
+
+  src = fetchzip {
+    url = "http://tmc.castleparadox.com/temp/fbc-${version}.tar.bz2";
```
```suggestion
    url = "https://tmc.castleparadox.com/temp/fbc-${version}.tar.bz2";
```

SuperSandro2000 commented on pkgs/development/compilers/fbc/default.nix:
```diff
@@ -0,0 +1,152 @@
+    "prefix=${placeholder "out"}"
+  ];
+
+  doCheck = (stdenv.buildPlatform == stdenv.hostPlatform);
```
```suggestion
  doCheck = stdenv.buildPlatform == stdenv.hostPlatform;
```

SuperSandro2000 commented on pkgs/development/compilers/fbc/default.nix:
```diff
@@ -0,0 +1,152 @@
+
+  enableParallelBuilding = true;
+
+  # Some tests seem to fail with too much (~64?) parallelism
```
Could we set an upper bound?

OPNA2608 replied:
We could but it'll require some testing on ofborg to find a good value, and making sense of the check output when all tests are run in parallel will be impossible… I'll see what I can do.

OPNA2608 replied:
Okay, seems to work.

SuperSandro2000 reviewed (commented):
I know to little about compiler...

OPNA2608 commented:
@ofborg build fbc

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1173

SuperSandro2000 commented on pkgs/development/compilers/fbc/default.nix:
```diff
@@ -0,0 +1,165 @@
+    "prefix=${placeholder "out"}"
+  ];
+
+  doCheck = stdenv.buildPlatform == stdenv.hostPlatform;
```
```suggestion
  doCheck = stdenv.buildPlatform.canExecute stdenv.hostPlatform;
```

OPNA2608 replied:
I tried `pkgsCross.gnu32.fbc` with this change + passing along `CC` to the test makefiles (otherwise they'd use the buildPlatform one) and one of the `log-tests` suite fails:
```
pretest/multi_module_ok.bmk : TEST_MODE=MULTI_MODULE_OK
cpp/call.bmk : TEST_MODE=MULTI_MODULE_OK
cpp/mangle.bmk : TEST_MODE=MULTI_MODULE_OK
cpp/fbcall.bmk : TEST_MODE=MULTI_MODULE_OK
cpp/bop.bmk : TEST_MODE=MULTI_MODULE_OK
cpp/derived.bmk : TEST_MODE=MULTI_MODULE_OK
make[2]: *** [log-tests.mk:259: cpp/derived.log] Error 1
```

`cpp/derived.log`:
```
cpp/derived.bmk : TEST_MODE=MULTI_MODULE_OK
make[3]: Entering directory '/build/source/tests'
/build/source/tests/../bin/fbc -i /build/source/tests/../inc -lang fb -target i686-unknown-linux-gnu -g -w 0  -m cpp/derived-fbc -c cpp/derived-fbc.bas -o cpp/derived-fbc.o
# Pass -m32 to get -m32 or -m64 as required
i686-unknown-linux-gnu-g++ -c -m32 -Wall -Wno-attributes -o cpp/derived-cpp.o cpp/derived-cpp.cpp
/build/source/tests/../bin/fbc -i /build/source/tests/../inc -target i686-unknown-linux-gnu -g cpp/derived-fbc.o cpp/derived-cpp.o -x cpp/derived-fbc
cpp/derived-fbc: error while loading shared libraries: libgcc_s.so.1: cannot open shared object file: No such file or directory
make[3]: *** [bmk-make.mk:137: run_test_ok] Error 1
make[3]: Leaving directory '/build/source/tests'
cpp/derived.bmk : RESULT=FAILED
```

```
↪ file /tmp/nix-build-fbc-i686-unknown-linux-gnu-1.09.0.drv-2/source/tests/cpp/derived-fbc
/tmp/nix-build-fbc-i686-unknown-linux-gnu-1.09.0.drv-2/source/tests/cpp/derived-fbc: ELF 32-bit LSB executable, Intel 80386, version 1 (SYSV), dynamically linked, interpreter /nix/store/yc911q1zcwaapdqn60fgzxmxnnh3dykc-glibc-i686-unknown-linux-gnu-2.35-163/lib/ld-linux.so.2, for GNU/Linux 2.6.32, not stripped

↪ nix-shell -p patchelf --run 'patchelf --print-rpath /tmp/nix-build-fbc-i686-unknown-linux-gnu-1.09.0.drv-2/source/tests/cpp/derived-fbc'
/nix/store/b5pm60096f94nh23c5xlph0690338cjj-fbc-i686-unknown-linux-gnu-1.09.0/lib64:/nix/store/b5pm60096f94nh23c5xlph0690338cjj-fbc-i686-unknown-linux-gnu-1.09.0/lib:/nix/store/3dvhlw5kbm3l8j6567l2n4h4db5hqg7k-i686-unknown-linux-gnu-stage-final-gcc-10.4.0-lib/i686-unknown-linux-gnu/lib:/nix/store/6crjr3xsx99fkg0hxcwhyawbrycj8k9j-ncurses-i686-unknown-linux-gnu-6.3-p20220507/lib:/nix/store/yc911q1zcwaapdqn60fgzxmxnnh3dykc-glibc-i686-unknown-linux-gnu-2.35-163/lib

↪ find /nix/store/3dvhlw5kbm3l8j6567l2n4h4db5hqg7k-i686-unknown-linux-gnu-stage-final-gcc-10.4.0-lib/ -name 'libgcc_s.so.1'
/nix/store/3dvhlw5kbm3l8j6567l2n4h4db5hqg7k-i686-unknown-linux-gnu-stage-final-gcc-10.4.0-lib/i686-unknown-linux-gnu/lib/libgcc_s.so.1

↪ strace /tmp/nix-build-fbc-i686-unknown-linux-gnu-1.09.0.drv-2/source/tests/cpp/derived-fbc 2>&1 | grep libgcc
openat(AT_FDCWD, "/nix/store/yc911q1zcwaapdqn60fgzxmxnnh3dykc-glibc-i686-unknown-linux-gnu-2.35-163/lib/libgcc_s.so.1", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/nix/store/yc911q1zcwaapdqn60fgzxmxnnh3dykc-glibc-i686-unknown-linux-gnu-2.35-163/lib/libgcc_s.so.1", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = -1 ENOENT (No such file or directory)
writev(2, [{iov_base="/tmp/nix-build-fbc-i686-unknown-"..., iov_len=83}, {iov_base=": ", iov_len=2}, {iov_base="error while loading shared libra"..., iov_len=36}, {iov_base=": ", iov_len=2}, {iov_base="libgcc_s.so.1", iov_len=13}, {iov_base=": ", iov_len=2}, {iov_base="cannot open shared object file", iov_len=30}, {iov_base=": ", iov_len=2}, {iov_base="No such file or directory", iov_len=25}, {iov_base="\n", iov_len=1}], 10/tmp/nix-build-fbc-i686-unknown-linux-gnu-1.09.0.drv-2/source/tests/cpp/derived-fbc: error while loading shared libraries: libgcc_s.so.1: cannot open shared object file: No such file or directory
```

Not sure what to make of this. `pkgsi686Linux.fbc` itself works though.

```
↪ file /tmp/nix-build-fbc-1.09.0.drv-0/source/tests/cpp/derived-fbc
/tmp/nix-build-fbc-1.09.0.drv-0/source/tests/cpp/derived-fbc: ELF 32-bit LSB executable, Intel 80386, version 1 (SYSV), dynamically linked, interpreter /nix/store/n2d7w74v3ixjsfd96b3207nzj6w4rl1f-glibc-2.35-163/lib/ld-linux.so.2, for GNU/Linux 2.6.32, not stripped

↪ nix-shell -p patchelf --run 'patchelf --print-rpath /tmp/nix-build-fbc-1.09.0.drv-0/source/tests/cpp/derived-fbc'
/nix/store/7kg5b400f3g5shkvkb647kwv5m2mnmb1-fbc-1.09.0/lib:/nix/store/dk0f8ffm2zyf4q3gqn035n13znhpdzmk-gcc-10.4.0-lib/lib:/nix/store/dfn02wyp7j76aw3f20mcbmar19siapzw-ncurses-6.3-p20220507/lib:/nix/store/n2d7w74v3ixjsfd96b3207nzj6w4rl1f-glibc-2.35-163/lib

↪ strace /tmp/nix-build-fbc-1.09.0.drv-0/source/tests/cpp/derived-fbc 2>&1 | grep libgcc
openat(AT_FDCWD, "/nix/store/dk0f8ffm2zyf4q3gqn035n13znhpdzmk-gcc-10.4.0-lib/lib/libgcc_s.so.1", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = 3
```

OPNA2608 replied:
I've added a comment explaining that tests do not pass under cross, I think that's all I can do here.

SuperSandro2000 replied:
> If that's the new recommended way of doing it then the manual should get updated.

Depends a bit on the usecase. It does not always work but eg when generating man pages it should be used and we should probably update the docs to consider this.

SuperSandro2000 commented on pkgs/development/compilers/fbc/default.nix:
```diff
@@ -0,0 +1,165 @@
+
+  doCheck = stdenv.buildPlatform == stdenv.hostPlatform;
+
+  checkTarget = lib.strings.concatStringsSep " " [
+    "unit-tests"
+    "warning-tests"
+    "log-tests"
+  ];
```
```suggestion
  checkTarget = "unit-tests warning-tests log-tests";
```
checkTarget is a string.

OPNA2608 commented:
Rebased to resolve merge conflict.

OPNA2608 commented:
Rebased again to resolve merge conflicts, replaced patch from a PR with the merged URL, squashed commits.

@SuperSandro2000 anything else to do here?

OPNA2608 committed:
fbc: init at 1.09.0 & 1.06-darwin-wip20160505
